apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/VolatileResourceManager.cs

```diff
@@ -27,28 +27,49 @@ internal sealed class VolatileResourceManager : ISinglePhaseNotification, ISpann
         private readonly SpannerConnection _spannerConnection;
         private readonly TimestampBound _timestampBound;
         private readonly TransactionId _transactionId;
-        private SpannerTransaction _transaction;
+        private Lazy<Task<SpannerTransaction>> _transaction;
+        private bool _hasExecutedDml;
 
         internal VolatileResourceManager(SpannerConnection spannerConnection, TimestampBound timestampBound, TransactionId transactionId)
         {
             _spannerConnection = spannerConnection;
             _timestampBound = timestampBound;
+            _transaction = new Lazy<Task<SpannerTransaction>>(CreateTransaction, LazyThreadSafetyMode.ExecutionAndPublication);
             _transactionId = transactionId;
         }
 
+        private SpannerTransaction SpannerTransaction { get => SpannerTransactionTask.Result; }
+
+        private Task<SpannerTransaction> SpannerTransactionTask
+        {
+            get => _transaction == null ? throw new ObjectDisposedException("The transaction has already been disposed") : _transaction.Value;
+        }
+
+        private bool IsTransactionCreated { get => _transaction != null && _transaction.IsValueCreated; }
+
+        private bool HasExecutedDmlOrMutations { get => IsTransactionCreated && (_hasExecutedDml || SpannerTransaction.HasMutations); }
+
         private Logger Logger => _spannerConnection.Logger;
 
+        private Task<SpannerTransaction> CreateTransaction()
+        {
+            return _timestampBound != null ? _spannerConnection.BeginReadOnlyTransactionAsync(_timestampBound)
+                : _transactionId != null ? Task.FromResult(_spannerConnection.BeginReadOnlyTransaction(_transactionId))
+                : _spannerConnection.BeginTransactionAsync();
+        }
+
         public void Dispose()
         {
             try
             {
-                _transaction?.Dispose();
-                // Protect against multiple disposals
+                if (_transaction != null && _transaction.IsValueCreated)
```

Comment:
You added a property for this:
```suggestion
                if (IsTransactionCreated)
```

Reply:
Done